### ISSUE TITLE
bug(StdCheats): dealERC721 fails for proxy contracts

### ISSUE THREAD

**Issue by SonOfMosiah** [foundry-rs/forge-std#318]
Getting the following error when using dealERC721 for UniswapV3 LP positions (i.e. NonfungiblePositionManager):
[FAIL. Reason: stdStorage find(StdStorage): Slot(s) not found.]

Seems find() is having a hard time getting the storage slot due to the implementation of EnumerableMaps.

Minimal reproducible repo [here](https://github.com/SonOfMosiah/forge-std_testDealERC721).

@wirew0lf

**Comment by mds1** [comment]
Probably because the NFT is a proxy contract, IIRC StdStorage does not support that currently and requires the slot to be in the `_target` address

**Comment by wirew0lf** [comment]
Indeed, the contract in the poc appears to be using a proxy:
https://etherscan.io/address/0xC36442b4a4522E871399CD717aBDD847Ab11FE88

I can think of a solution for this type of cases, although not ideal but it would be the only way to fix this without major changes to StdStorage. Basically it'd be pranking the owner and performing a `transferFrom`. This is not ideal, mainly because:
- If transfers are somehow disabled in the contract such as some debt or position tokens the deal will fail.
- If this is triggered inside a `startPrank` the first prank will be stopped as I think there is no way to perform a prank inside antoher one, or resume it after the deal prank is performed.

So this could be implemented as a backup method. But I'm curious to hear your thoughts and opinions on this.

**Comment by mds1** [comment]
Ideally we could add proxy support to StdStorage, though I'm not sure offhand how much work that would be. Might not be too bad if we only support standard proxy implementations like EIP-1967 as a fallback if a storage slot couldn't be found normally

**Comment by wirew0lf** [comment]
> Ideally we could add proxy support to StdStorage, though I'm not sure offhand how much work that would be. Might not be too bad if we only support standard proxy implementations like EIP-1967 as a fallback if a storage slot couldn't be found normally

This would be great definitely. I'm not too familiar with how StdStorage works internally, but I can take a look and try to work on this when I have some spare time.

**Comment by nine-december** [comment]
+1 here. Same issue trying to deal tokens behind a proxy-implementation pattern.

**Comment by PutraLaksmana** [comment]
usually when I deal proxy tokens, 
I make it like this:
```
IERC20 tokenproxy = IERC20("address");
address whale = address("address_holder_tokenproxy");

function setUp() public {
        vm.createSelectFork("mainnet", block.number);
        vm.startPrank(whale);
        tokenproxy.transfer(address(this), 10000e18);
        vm.stopPrank();
    }
```
It's worked for me.

**Comment by mds1** [comment]
Mainnet USDC recently ugpraded and now stored balance in a packed slot, so `deal` for it will be fixed once https://github.com/foundry-rs/forge-std/pull/505 is merged

**Comment by kamuik16** [comment]
> Mainnet USDC recently ugpraded and now stored balance in a packed slot, so `deal` for it will be fixed once #505 is merged

Hey! I am having the same issue! I am on the latest foundry version. Still getting an error like this while dealing Mainnet USDC to an address.

```bash
[FAIL. Reason: revert: stdStorage find(StdStorage): Slot(s) not found.]
```
Any solution for this?

**Comment by mds1** [comment]
Correct, you will need to upgrade your forge-std dependency to the latest release, not just your foundry version

**Comment by rpedroni** [comment]
Can confirm updating `forge-std` works when `deal`ing proxy USDC
